File: src/Layout/container/bicontainer/BiContainer.java

package Layout.container.bicontainer;

import java.util.HashMap;
import java.util.Map;

import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.geometry.Orientation;
import javafx.scene.Node;

import Layout.Areas.Splitter;
import Layout.Component;
import Layout.container.Container;

import static javafx.geometry.Orientation.VERTICAL;

/**
 * Implementation of {@link Container Container} containing two children.
 * <p>
 * @author uranium
 *
 * Warning: do not use this class.
 * @TODO implement load() properly, currently works only for Containers. Hence
 * the abstract class type to avoid misuse. Do not use this class as non-pure
 * Container. See addChild and the exception.
 */
public class BiContainer extends Container<Splitter> {

    /** Orientation of this container. */
    public final ObjectProperty<Orientation> orientation = new SimpleObjectProperty<>(VERTICAL);
    private final Map<Integer, Component> children = new HashMap<>();


    public BiContainer(Orientation o) {
        orientation.set(o);
    }

    @Override
    public Node load() {
        // lazy load (needed because of the serialization ommiting this field)
        if (ui == null) ui = new Splitter(this);

        ui.setChild1(children.get(1));
        ui.setChild2(children.get(2));

        return ui.getRoot();
    }

    /**
     * @return the children
     */
    @Override
    public Map<Integer, Component> getChildren() {
        return children;
    }

    /**
     * { @inheritDoc }
     * Index can only take on value 1 or 2. Other values will do
     * nothing.
     * @throws UnsupportedOperationException if component not Container type.
     */
    @Override
    public void addChild(Integer index, Component c) {
        if(index == null) return;
        if (index<1 || index>2)
            throw new IndexOutOfBoundsException("Index " + index + " not supported. Only null,1,2 values supported.");

        if(c==null) children.remove(index);
        else children.put(index, c);

        if(ui!=null) ui.setComponent(index, c);
        setParentRec();
    }

    public void switchCildren() {
        Component c1 = children.get(1);
        Component c2 = children.get(2);
        children.clear();
        children.put(1, c2);
        children.put(2, c1);
        load();
    }

    @Override
    public Integer getEmptySpot() {
        if(children.get(1)==null) return 1;
        if(children.get(2)==null) return 2;
        else return null;
    }

    @Override
    public void show() {
//        super.show();
        if(ui!=null) ui.show();
    }

    @Override
    public void hide() {
//        super.hide();
//        if(gui !=null) gui.widgets.values().forEach(Area::hide);
                if(ui!=null) ui.hide();
    }

}

File: src/player/main/sp/it/pl/layout/container/BiContainer.kt
package sp.it.pl.layout.container

import javafx.geometry.Orientation
import javafx.geometry.Orientation.VERTICAL
import javafx.geometry.Orientation.HORIZONTAL
import javafx.scene.Node
import sp.it.pl.layout.BiContainerDb
import sp.it.pl.layout.Component
import sp.it.util.access.v
import sp.it.util.collections.filterNotNullValues
import sp.it.util.dev.failIf
import java.util.HashMap

/** [Container] containing two children split vertically or horizontally. */
class BiContainer: Container<BiContainerUi> {

   override val name = "BiContainer"
   /** Orientation of this container. */
   val orientation = v(VERTICAL)
   /**
    * Position of the divider.
    * * [0,1] range if [absoluteSize] 0
    * * [0, Infinity) range of width ([orientation] == [HORIZONTAL]) or height ([orientation] == [VERTICAL]) of the absolute child if [absoluteSize] 1 or 2.
    *
    * This value goes hand in hand with [absoluteSize], which is the dominant property in this relationship. Therefore:
    * * This value must be set to value required by [absoluteSize].
    * * This value converts itself to the other value range when [absoluteSize] changes.
    * * Initial value (before the component is displayed) will not be converted and should be provided in correct range.
    *   This is because the conversion depends on the actual ui size
    */
   val position = v(0.5)
   /** Whether child is resized absolutely (retains its size on layout). 0 == none, 1 == child 1, 2 == child 2 */
   val absoluteSize = v(0)
   /** Whether child is hidden so the other covers the entire space. 0 == none, -1 == child 1, 1 == child 2 */
   val collapsed = v(0)
   /** Whether the children should appear as one, i.e., the divider has little to no visibility */
   val joined = v(false)
   private val children = HashMap<Int, Component>()

   constructor(state: BiContainerDb = BiContainerDb()): super(state) {
      orientation.value = state.orientation
      position.value = state.position
      absoluteSize.value = state.absoluteSize
      collapsed.value = state.collapsed
      joined.value = state.joined
      children += state.children.filter { it.key in 1..2 }.mapValues { it.value?.toDomain() }.filterNotNullValues()
      setChildrenParents()
   }

   constructor(o: Orientation): this(BiContainerDb(orientation = o))

   override fun load(): Node {
      ui = ui ?: BiContainerUi(this)

      ui.setComponent(1, children[1])
      ui.setComponent(2, children[2])

      return ui.root
   }

   override fun getChildren() = children

   /**
    * @param index null does nothing, value other than 1 or 2 throws exception
    */
   override fun addChild(index: Int?, c: Component?) {
      if (index==null) return
      failIf(index!=1 && index!=2) { "Index=$index must be 1 or 2" }

      if (c==null) children.remove(index)
      else children[index] = c

      ui?.setComponent(index, c)
      setParentRec()
   }

   fun swapChildren() {
      val c1 = children[1]
      val c2 = children[2]
      when {
         c1==null && c2==null -> return
         c1==null && c2!=null -> c2.swapWith(this, 1)
         c1!=null && c2==null || c1!=null && c2!=null -> c1.swapWith(this, 2)
      }
   }

   override fun getEmptySpot() = when {
      children[1]==null -> 1
      children[2]==null -> 2
      else -> null
   }

   override fun show() = ui?.show() ?: Unit

   override fun hide() = ui?.hide() ?: Unit

   override fun toDb() = BiContainerDb(id, orientation.value, position.value, absoluteSize.value, collapsed.value, joined.value, loadType.value, locked.value, children.mapValues { it.value.toDb() }, properties)

}